neo/Network/P2P/Payloads/WitnessScope.cs

```diff
@@ -0,0 +1,38 @@
+﻿namespace Neo.Network.P2P.Payloads
+{
+    public enum WitnessScope : byte
+    {
+        /// <summary>
+        /// (neo2) - no params
+        /// </summary>
+        Global = 0x00,
+
+        /// <summary>
+        /// EntryOnly means that this condition must hold: EntryScriptHash == CallingScriptHash
+        /// No params is needed, as the witness/permission/signature given on first invocation will automatically expire if entering deeper internal invokes
+        /// This can be default safe choice for native NEO/GAS (previously used on Neo 2 as "attach" mode)
+        /// </summary>
+        EntryOnly = 0x01,
+
+        /// <summary>
+        /// Custom hash for contract-specific
+        /// </summary>
+        CustomScriptHash = 0x02,
+
+        // 0x03 -> composition between EntryOnly and CustomScriptHash. 
+        // Example: we invoke NEO native transfer using 0x01 (it should work).
+        // We invoke NEO native transfer using 0x03, and custom hash for GAS asset (it should fail)
+
+        /// <summary>
+        ///  Custom pubkey for group members
+        /// </summary>
+        ExecutingGroupPubKey = 0x04
```

Comment:
How does it work?

Reply:
The idea is to only allow the scope inside this group. I haven't added tests to it yet.
One thing I don't know yet is if it's possible to get to another contract, when you are in a group. If a group forbids you from "getting out", then this scope would endup becoming the same as Global (I suspect that).

Reply:
Commit `filtering groups by pubkey` will make this clear for you Erik, I think.

Reply:
https://github.com/neo-project/neo/pull/968/commits/862ff4b7b1751b5a3a54697008a225d4422ca3e9#diff-3425c2ccbc7797fa6cdc01a9a753565bR178-R184